impeller/geometry/path.h

```diff
@@ -75,107 +77,261 @@ class Path {
+    };
+
+    size_t GetComponentCount(std::optional<ComponentType> type = {}) const;
+
+    void SetFillType(FillType fill);
+
+    FillType GetFillType() const;
+
+    bool IsConvex() const;
+
+    void AddLinearComponent(Point p1, Point p2);
+
+    void AddQuadraticComponent(Point p1, Point cp, Point p2);
+
+    void AddCubicComponent(Point p1, Point cp1, Point cp2, Point p2);
+
+    void AddContourComponent(Point destination, bool is_closed = false);
+
+    void SetContourClosed(bool is_closed);
+
+    template <class T>
+    using Applier = std::function<void(size_t index, const T& component)>;
+    void EnumerateComponents(
+        const Applier<LinearPathComponent>& linear_applier,
+        const Applier<QuadraticPathComponent>& quad_applier,
+        const Applier<CubicPathComponent>& cubic_applier,
+        const Applier<ContourComponent>& contour_applier) const;
+
+    bool GetLinearComponentAtIndex(size_t index,
+                                   LinearPathComponent& linear) const;
+
+    bool GetQuadraticComponentAtIndex(size_t index,
+                                      QuadraticPathComponent& quadratic) const;
+
+    bool GetCubicComponentAtIndex(size_t index,
+                                  CubicPathComponent& cubic) const;
+
+    bool GetContourComponentAtIndex(size_t index,
+                                    ContourComponent& contour) const;
+
+    bool UpdateLinearComponentAtIndex(size_t index,
+                                      const LinearPathComponent& linear);
+
+    bool UpdateQuadraticComponentAtIndex(
+        size_t index,
+        const QuadraticPathComponent& quadratic);
+
+    bool UpdateCubicComponentAtIndex(size_t index, CubicPathComponent& cubic);
+
+    bool UpdateContourComponentAtIndex(size_t index,
+                                       const ContourComponent& contour);
+
+    Polyline CreatePolyline(Scalar scale) const;
+
+    std::optional<Rect> GetBoundingBox() const;
+
+    std::optional<Rect> GetTransformedBoundingBox(
+        const Matrix& transform) const;
+
+    std::optional<std::pair<Point, Point>> GetMinMaxCoveragePoints() const;
+
+    void SetConvexity(Convexity value);
 
-  struct ComponentIndexPair {
-    ComponentType type = ComponentType::kLinear;
-    size_t index = 0;
+    FillType fill_ = FillType::kNonZero;
+    Convexity convexity_ = Convexity::kUnknown;
+    std::vector<ComponentIndexPair> components_;
+    std::vector<LinearPathComponent> linears_;
+    std::vector<QuadraticPathComponent> quads_;
+    std::vector<CubicPathComponent> cubics_;
+    std::vector<ContourComponent> contours_;
 
-    ComponentIndexPair() {}
+    void InvalidateCache() {
+      cached_bounding_box_.reset();
+      cached_polyline_.reset();
+    }
 
-    ComponentIndexPair(ComponentType a_type, size_t a_index)
-        : type(a_type), index(a_index) {}
+    mutable std::optional<std::optional<Rect>> cached_bounding_box_;
+    mutable std::optional<Polyline> cached_polyline_;
```

Comment:
This is a separate cache from the tessellation cache, right?

Reply:
Path instance caches polyline and bounding box. Polyline instance caches tessellation data.